flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java

```diff
@@ -0,0 +1,680 @@
+
+		return pendingRequest.getSlotFuture();
+	}
+
+	private void declareResourceRequirements(Collection<ResourceRequirement> resourceRequirements) {
+		assertRunningInMainThread();
+
+		LOG.debug("Declare new resource requirements for job {}: {}.", jobId, resourceRequirements);
+
+		declareResourceRequirementServiceConnectionManager.declareResourceRequirements(ResourceRequirements.create(jobId, jobManagerAddress, resourceRequirements));
+	}
+
+	@Override
+	public Optional<ResourceID> failAllocation(AllocationID allocationID, Exception cause) {
+		throw new UnsupportedOperationException("Please call failAllocation(ResourceID, AllocationID, Exception)");
+	}
+
+	@Override
+	public Optional<ResourceID> failAllocation(@Nullable ResourceID resourceId, AllocationID allocationID, Exception cause) {
+		assertRunningInMainThread();
+
+		Preconditions.checkNotNull(allocationID);
+		Preconditions.checkNotNull(resourceId, "This slot pool only supports failAllocation calls coming from the TaskExecutor.");
+
+		ResourceCounter previouslyFulfilledRequirements = declarativeSlotPool.releaseSlot(allocationID, cause);
+		if (!previouslyFulfilledRequirements.isEmpty()) {
+			declarativeSlotPool.decreaseResourceRequirementsBy(previouslyFulfilledRequirements);
+		}
+
+		if (declarativeSlotPool.containsSlots(resourceId)) {
+			return Optional.empty();
+		} else {
+			return Optional.of(resourceId);
+		}
+	}
+
+	@Override
+	public void releaseSlot(@Nonnull SlotRequestId slotRequestId, @Nullable Throwable cause) {
+		LOG.debug("Release slot with slot request id {}", slotRequestId);
+		assertRunningInMainThread();
+
+		final PendingRequest pendingRequest = pendingRequests.remove(slotRequestId);
+
+		if (pendingRequest != null) {
+			declarativeSlotPool.decreaseResourceRequirementsBy(ResourceCounter.withResource(pendingRequest.getResourceProfile(), 1));
+			pendingRequest.failRequest(new FlinkException(
+					String.format("Pending slot request with %s has been released.", pendingRequest.getSlotRequestId()),
+					cause));
+		} else {
+			final AllocationID allocationId = fulfilledRequests.remove(slotRequestId);
+
+			if (allocationId != null) {
+				ResourceCounter previouslyFulfilledRequirement = declarativeSlotPool.freeReservedSlot(allocationId, cause, clock.relativeTimeMillis());
+				if (!previouslyFulfilledRequirement.isEmpty()) {
+					declarativeSlotPool.decreaseResourceRequirementsBy(previouslyFulfilledRequirement);
+				}
+			} else {
+				LOG.debug("Could not find slot which has fulfilled slot request {}. Ignoring the release operation.", slotRequestId);
+			}
+		}
+	}
+
+	@Override
+	public void notifyNotEnoughResourcesAvailable(Collection<ResourceRequirement> acquiredResources) {
+		assertRunningInMainThread();
+
+		componentMainThreadExecutor.schedule(
+				failPendingRequests(),
+				resourceAcquisitionTimeout.toMillis(),
+				TimeUnit.MILLISECONDS);
```

Comment:
Why are we scheduling the `failPendingRequests` here and don't execute it immediately?

Reply:
IIRC we wanted to have a timeout here in particular for a newly started cluster to ensure the job can run even if the task executors are starting up a bit slowly.

Reply:
no wait, that case is covered by the resource managers startup period...

I don't quite remember then. It does use the slot request timeout as a default, but then again we already setup a timeout within `requestNewAllocatedSlot`. But this will likely just result in the allocation being retried?

I guess that is it; in order to differentiate between timeouts due to network issues, and not enough resources being available......................but then why not fail immediately?

OH. I remember now; we actually discussed this once. My thinking back then was that we should not fail immediately because we should honor the slot request timeout, but you mentioned that this is just a maximum, and if a component has more information available then it is free to fail earlier.

So this is just something I forgot to change :/

Reply:
Sounds good to me.